path_login.go

```diff
@@ -0,0 +1,239 @@
+			logical.UpdateOperation: b.pathLogin,
+		},
+
+		HelpSynopsis:    pathLoginHelpSyn,
+		HelpDescription: pathLoginHelpDesc,
+	}
+}
+
+func (b *jwtAuthBackend) pathLogin(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	token := d.Get("token").(string)
+	if len(token) == 0 {
+		return logical.ErrorResponse("missing token"), nil
+	}
+
+	roleName := d.Get("role").(string)
+	if len(roleName) == 0 {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse("role could not be found"), nil
+	}
+
+	if req.Connection != nil && !cidrutil.RemoteAddrIsOk(req.Connection.RemoteAddr, role.BoundCIDRs) {
+		return logical.ErrorResponse("request originated from invalid CIDR"), nil
+	}
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	// Here is where things diverge. If it is using OIDC discovery, validate
+	// that way; otherwise validate against the locally configured keys. Once
+	// things are validated, we re-unify the request path when evaluating the
+	// claims.
+	allClaims := map[string]interface{}{}
+	switch {
+	case len(config.ParsedJWTPubKeys) != 0:
+		parsedJWT, err := jwt.ParseSigned(token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error parsing token: {{err}}", err).Error()), nil
+		}
+
+		claims := jwt.Claims{}
+
+		var valid bool
+		for _, key := range config.ParsedJWTPubKeys {
+			if err := parsedJWT.Claims(key, &claims, allClaims); err == nil {
+				valid = true
+				break
+			}
+		}
+		if !valid {
+			return logical.ErrorResponse("no known key successfully validated the token signature"), nil
+		}
+
+		expected := jwt.Expected{
+			Issuer:   config.BoundIssuer,
+			Subject:  role.BoundSubject,
+			ID:       role.BoundID,
+			Audience: jwt.Audience(role.BoundAudiences),
+			Time:     time.Now(),
+		}
+
+		if err := claims.Validate(expected); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
+		}
+
+	case config.OIDCIssuerURL != "":
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		verifier := provider.Verifier(&oidc.Config{
+			SkipClientIDCheck: true,
+		})
+
+		idToken, err := verifier.Verify(ctx, token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating signature: {{err}}", err).Error()), nil
+		}
+
+		if err := idToken.Claims(allClaims); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("unable to successfully parse all claims from token: {{err}}", err).Error()), nil
+		}
+
+		var jtiVal string
+		if jti, ok := allClaims["jti"]; ok {
```

Comment:
My reading of the jti claim is that it is not an id but more or less a nonce.  https://tools.ietf.org/html/rfc7519#section-4.1.7

Reply:
I guess my point here is that if the value is never to be reused, it probably has limited value as a bound parameter.

Reply:
Yep, thanks for the pointer. I think you're right. I'll take it out. I had thought it was an ability to scope similar to `sub` but it looks like it is not.